report_thesis/src/sections/proposed_approach/optimization_framework.tex

```diff
@@ -0,0 +1,193 @@
+We believe that this approach will allow us to best identify the top-$n$ pipelines to be used in our stacking ensemble.
+
+To manage the optimization process the function receives the number of trials to run, a list of models, and a list of oxides, seen in line~\ref{step:initialize_run_process}, and initializes the optimization environment by setting up the sampler, pruner, and the objective function, seen in lines~\ref{step:initialize_sampler} to~\ref{step:initialize_objective}.
+
+The sampler is responsible for managing the search space of the hyperparameters for the optimization process.
+This means that any hyperparameters being evaluated, for any preprocessor or model, will be managed by this sampler, which allows us to optimize for all hyperparameters at the same time.
+Optuna provides several options for samplers that have different characteristics and each have their strengths and weaknesses.
+However, because we require multi-objective optimization, this naturally limits the choice of sampler to those that support this.
+For our framework we chose to use the Tree-structured Parzen Estimator (TPE) sampler, due to its stated optimization efficiency and its ability to handle all use cases. 
+Additionally, the TPE sampler allows us to control how many of the trials to be reserved for exploration, which is beneficial when the search space is large. \cite{optuna_2019}
+
+The pruner is responsible for stopping unpromising trials early, to reduce the computational costs of the optimization process and is initialized in line~\ref{step:initialize_pruner}.
+Similar to the sampler, there exist several options for the choice of pruner.
+In our case we chose to use the Hyperband pruner, due to its balanced approach when pruning.
+Initially, the pruner will consider more configurations, but allow them to run for a shorter time.
+Through several iterations, it will gradually reduce the number of configurations being considered, but allow them to run for a longer time.
+This provides a higher likelihood that promising configurations are not pruned too early, while still reducing the computational costs.\cite{optuna_2019}
+
+Guiding both the sampler and the pruner is the objective function, which evaluates the performance of each trial, seeking to either minimize or maximize some specified metrics.
+In our case we are seeking to minimize the \texttt{rmse\_cv} and \texttt{std\_dev\_cv}, defined previously.
+This can be seen in line~\ref{step:initialize_objective}.
+
+The role of the objective function is to provide the metric data that the sampler and pruner need to make informed decisions.
+As we step through each oxide and model in Lines~\ref{step:oxide_loop} to~\ref{step:model_loop}, we call the $objective$ function with the $Optimize\_Combined\_Objective$ function, seen in Line~\ref{step:optimize_combined_objective}.
```

Comment:
I still think the naming of these functions is confusing.
You say that:
> The role of the objective function is to provide the metric data that the sampler and pruner need to make informed decisions.

But the `objective` function takes the function that actually returns those metrics?
![image](https://github.com/chhoumann/thesis-chemcam/assets/29108628/d7a3eab8-c615-44e9-be1f-7eea496f9ffd)


And another thing in the image: you pass `o` for the current oxide to `Optimize_Combined_Objective`, yet it doesn't state that as a parameter.
And what's with the Pascal_Snake_Case / snake_case mix?

I think your algorithm setup would be basically perfect if:
- you supported `o` as a parameter
- you fix the issue I've mentioned for the millionth time now: 
![image](https://github.com/chhoumann/thesis-chemcam/assets/29108628/664d121f-593e-4eff-9ee9-eed9b850871d)
- you fix the naming of `objective` so it's more like the `optimizer` or something, and you rename `Optimize_Combined_Objective` as `objective`, because that's literally what it is & behaves as (except for the loop). You're basically mixing optimization and scoring.
- you move the `for each trial` thing _out_ of the objective function! It doesn't make sense - I think it belongs with the optimizer

So basically, move all the optimization stuff to Run Optimization Process, and keep the scoring stuff in Objective.

And a minor point: the pruner part of the algorithm is slightly confusing, but also not all bad. If it isn't much effort to make clearer, do so, but don't spend much time on it.